synapse/handlers/oidc_handler.py

```diff
@@ -0,0 +1,452 @@
+
+        if "error" not in resp:
+            return resp
+
+        error = resp["error"]
+        description = resp.get("error_description", error)
+        raise OidcError(error, description)
+
+    async def _fetch_userinfo(self, token: Dict[str, str]) -> UserInfo:
+        metadata = await self.load_metadata()
+
+        resp = await self._http_client.get_json(
+            metadata["userinfo_endpoint"],
+            headers={"Authorization": ["Bearer {}".format(token["access_token"])]},
+        )
+
+        return UserInfo(resp)
+
+    async def _parse_id_token(self, token: Dict[str, str], nonce: str) -> UserInfo:
+        """Return an instance of UserInfo from token's ``id_token``."""
+        metadata = await self.load_metadata()
+        claims_params = {
+            "nonce": nonce,
+            "client_id": self._client_auth.client_id,
+        }
+        if "access_token" in token:
+            claims_params["access_token"] = token["access_token"]
+            claims_cls = CodeIDToken
+        else:
+            claims_cls = ImplicitIDToken
+
+        alg_values = metadata.get("id_token_signing_alg_values_supported", ["RS256"])
+
+        jwt = JsonWebToken(alg_values)
+
+        claim_options = {"iss": {"values": [metadata["issuer"]]}}
+
+        # Try to decode the keys in cache first, then retry by forcing the keys
+        # to be reloaded
+        jwk_set = await self.load_jwks()
+        try:
+            claims = jwt.decode(
+                token["id_token"],
+                key=jwk_set,
+                claims_cls=claims_cls,
+                claims_options=claim_options,
+                claims_params=claims_params,
+            )
+        except ValueError:
+            jwk_set = await self.load_jwks(force=True)  # try reloading the jwks
+            claims = jwt.decode(
+                token["id_token"],
+                key=jwk_set,
+                claims_cls=claims_cls,
+                claims_options=claim_options,
+                claims_params=claims_params,
+            )
+
+        claims.validate(leeway=120)  # allows 2 min of clock skew
+        return UserInfo(claims)
+
+    async def handle_redirect_request(self, request, client_redirect_url: bytes) -> str:
+        """Handle an incoming request to /login/sso/redirect
+
+        Args:
+            request (SynapseRequest)
+            client_redirect_url (bytes): the URL that we should redirect the
+                client to when everything is done
+
+        Returns:
+            str: URL to redirect to
+        """
+
+        state = generate_token()
+        nonce = generate_token()
+
+        cookie = self._macaroon_generator.generate_oidc_session_token(
+            state=state, nonce=nonce, client_redirect_url=client_redirect_url.decode(),
+        )
+        request.addCookie(
+            SESSION_COOKIE_NAME,
+            cookie,
+            path="/_synapse/oidc",
+            max_age="3600",
+            httpOnly=True,
```

Comment:
Do we want to set `sameSite` or `secure` here?

Reply:
SameSite probably, as older browsers should just ignore it, right? But secure only if served over HTTPS. I think lots of debugging scenarios may have stuff running over plain HTTP

Reply:
I think I came to the same conclusions when thinking about it briefly.

Reply:
I had issues with SameSite if set to strict, as the cookie would not be in the callback, as the browser does not send SameSite cookies when redirecting from another host

Reply:
Yes, I think it should be Lax, not Strict and absolutely not None